compiler/rustc_ast/src/tokenstream.rs

```diff
@@ -296,6 +303,96 @@ pub struct AttrsTarget {
 #[derive(Clone, Debug, Default, Encodable, Decodable)]
 pub struct TokenStream(pub(crate) Lrc<Vec<TokenTree>>);
 
+struct HashEncoder<H: std::hash::Hasher> {
+    hasher: H,
+}
+
+impl<H: std::hash::Hasher> Encoder for HashEncoder<H> {
+    fn emit_usize(&mut self, v: usize) {
+        self.hasher.write_usize(v)
+    }
+
+    fn emit_u128(&mut self, v: u128) {
+        self.hasher.write_u128(v)
+    }
+
+    fn emit_u64(&mut self, v: u64) {
+        self.hasher.write_u64(v)
+    }
+
+    fn emit_u32(&mut self, v: u32) {
+        self.hasher.write_u32(v)
+    }
+
+    fn emit_u16(&mut self, v: u16) {
+        self.hasher.write_u16(v)
+    }
+
+    fn emit_u8(&mut self, v: u8) {
+        self.hasher.write_u8(v)
+    }
+
+    fn emit_isize(&mut self, v: isize) {
+        self.hasher.write_isize(v)
+    }
+
+    fn emit_i128(&mut self, v: i128) {
+        self.hasher.write_i128(v)
+    }
+
+    fn emit_i64(&mut self, v: i64) {
+        self.hasher.write_i64(v)
+    }
+
+    fn emit_i32(&mut self, v: i32) {
+        self.hasher.write_i32(v)
+    }
+
+    fn emit_i16(&mut self, v: i16) {
+        self.hasher.write_i16(v)
+    }
+
+    fn emit_raw_bytes(&mut self, s: &[u8]) {
+        self.hasher.write(s)
+    }
+}
+
+impl<H: std::hash::Hasher> SpanEncoder for HashEncoder<H> {
+    fn encode_span(&mut self, span: Span) {
+        span.hash(&mut self.hasher)
+    }
+
+    fn encode_symbol(&mut self, symbol: Symbol) {
+        symbol.hash(&mut self.hasher)
+    }
+
+    fn encode_expn_id(&mut self, expn_id: rustc_span::ExpnId) {
+        expn_id.hash(&mut self.hasher)
+    }
+
+    fn encode_syntax_context(&mut self, syntax_context: rustc_span::SyntaxContext) {
+        syntax_context.hash(&mut self.hasher)
+    }
+
+    fn encode_crate_num(&mut self, crate_num: CrateNum) {
+        crate_num.hash(&mut self.hasher)
+    }
+
+    fn encode_def_index(&mut self, def_index: DefIndex) {
+        def_index.hash(&mut self.hasher)
+    }
+
+    fn encode_def_id(&mut self, def_id: rustc_span::def_id::DefId) {
+        def_id.hash(&mut self.hasher)
+    }
+}
+
+impl Hash for TokenStream {
```

Comment:
Tried removing the `Hash`-impl again, but it seems the query system requires it somehow:

```
   Compiling rustc_middle v0.0.0 (/home/felix/gits/rust/compiler/rustc_middle)
error[E0277]: the trait bound `rustc_ast::tokenstream::TokenStream: Hash` is not satisfied
    --> compiler/rustc_middle/src/query/plumbing.rs:434:28
     |
275  |  / macro_rules! define_callbacks {
276  |  |     (
277  |  |      $($(#[$attr:meta])*
278  |  |         [$($modifiers:tt)*] fn $name:ident($($K:tt)*) -> $V:ty,)*) => {
...     |
434  |  |                 pub $name: DynamicQuery<'tcx, queries::$name::Storage<'tcx>>,
     |  |                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Hash` is not implemented for `rustc_ast::tokenstream::TokenStream`, which is required by `DefaultCache<(rustc_span::LocalExpnId, rustc_data_structures::svh::Svh, &'tcx rustc_ast::tokenstream::TokenStream), Erased<query::erase::Erase<Result<&'tcx rustc_ast::tokenstream::TokenStream, ()>>::{opaque#0}>>: rustc_query_system::query::QueryCache`
...     |
490  |  |     };
491  |  | }
     |  |_- in this expansion of `define_callbacks!` (#2)
     |
    ::: compiler/rustc_middle/src/query/mod.rs:108:1
     |
108  | /  rustc_queries! {
109  | |      query derive_macro_expansion(key: (LocalExpnId, Svh, &'tcx TokenStream)) -> Result<&'tcx TokenStream, ()> {
110  | |          // eval_always
111  | |          // no_hash
...    |
2335 | |      }
2336 | |  }
     | |  -
     | |  |
     | |__in this expansion of `rustc_query_append!` (#1)
     |    in this macro invocation (#2)
2337 |
2338 |    rustc_query_append! { define_callbacks! }
     |    ----------------------------------------- in this macro invocation (#1)
     |
     = help: the trait `rustc_query_system::query::QueryCache` is implemented for `DefaultCache<K, V>`
     = note: required for `&'tcx rustc_ast::tokenstream::TokenStream` to implement `Hash`
     = note: 1 redundant requirement hidden
     = note: required for `(rustc_span::LocalExpnId, rustc_data_structures::svh::Svh, &'tcx rustc_ast::tokenstream::TokenStream)` to implement `Hash`
     = note: required for `DefaultCache<(LocalExpnId, Svh, &TokenStream), Erased<Erase<Result<&TokenStream, ()>>::{opaque#0}>>` to implement `rustc_query_system::query::QueryCache`
note: required by a bound in `DynamicQuery`
    --> compiler/rustc_middle/src/query/plumbing.rs:22:34
     |
22   | pub struct DynamicQuery<'tcx, C: QueryCache> {
     |                                  ^^^^^^^^^^ required by this bound in `DynamicQuery`
     = note: the full name for the type has been written to '/home/felix/gits/rust/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/rustc_middle-3afd8e69956fb57f.long-type-16771981020465904745.txt'
     = note: consider using `--verbose` to print the full type name to the console
```

Adding `no_hash` also didn't help (i.e., error stayed the same). I know there is something like `arena_cache`, but I'm not really familiar with the query/caching system, so no idea if that would be better here/not require the `Hash`-impl.

<details><summary>Full error message (long!)</summary>
<pre>

   Compiling rustc_middle v0.0.0 (/home/felix/gits/rust/compiler/rustc_middle)
error[E0277]: the trait bound `rustc_ast::tokenstream::TokenStream: Hash` is not satisfied
    --> compiler/rustc_middle/src/query/plumbing.rs:434:28
     |
275  |  / macro_rules! define_callbacks {
276  |  |     (
277  |  |      $($(#[$attr:meta])*
278  |  |         [$($modifiers:tt)*] fn $name:ident($($K:tt)*) -> $V:ty,)*) => {
...     |
434  |  |                 pub $name: DynamicQuery<'tcx, queries::$name::Storage<'tcx>>,
     |  |                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Hash` is not implemented for `rustc_ast::tokenstream::TokenStream`, which is required by `DefaultCache<(rustc_span::LocalExpnId, rustc_data_structures::svh::Svh, &'tcx rustc_ast::tokenstream::TokenStream), Erased<query::erase::Erase<Result<&'tcx rustc_ast::tokenstream::TokenStream, ()>>::{opaque#0}>>: rustc_query_system::query::QueryCache`
...     |
490  |  |     };
491  |  | }
     |  |_- in this expansion of `define_callbacks!` (#2)
     |
    ::: compiler/rustc_middle/src/query/mod.rs:108:1
     |
108  | /  rustc_queries! {
109  | |      query derive_macro_expansion(key: (LocalExpnId, Svh, &'tcx TokenStream)) -> Result<&'tcx TokenStream, ()> {
110  | |          // eval_always
111  | |          // no_hash
...    |
2335 | |      }
2336 | |  }
     | |  -
     | |  |
     | |__in this expansion of `rustc_query_append!` (#1)
     |    in this macro invocation (#2)
2337 |
2338 |    rustc_query_append! { define_callbacks! }
     |    ----------------------------------------- in this macro invocation (#1)
     |
     = help: the trait `rustc_query_system::query::QueryCache` is implemented for `DefaultCache<K, V>`
     = note: required for `&'tcx rustc_ast::tokenstream::TokenStream` to implement `Hash`
     = note: 1 redundant requirement hidden
     = note: required for `(rustc_span::LocalExpnId, rustc_data_structures::svh::Svh, &'tcx rustc_ast::tokenstream::TokenStream)` to implement `Hash`
     = note: required for `DefaultCache<(LocalExpnId, Svh, &TokenStream), Erased<Erase<Result<&TokenStream, ()>>::{opaque#0}>>` to implement `rustc_query_system::query::QueryCache`
note: required by a bound in `DynamicQuery`
    --> compiler/rustc_middle/src/query/plumbing.rs:22:34
     |
22   | pub struct DynamicQuery<'tcx, C: QueryCache> {
     |                                  ^^^^^^^^^^ required by this bound in `DynamicQuery`
     = note: the full name for the type has been written to '/home/felix/gits/rust/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/rustc_middle-3afd8e69956fb57f.long-type-16771981020465904745.txt'
     = note: consider using `--verbose` to print the full type name to the console

error[E0277]: the trait bound `rustc_ast::tokenstream::TokenStream: Hash` is not satisfied
    --> compiler/rustc_middle/src/ty/context.rs:1267:29
     |
1267 |     sync::assert_dyn_sync::<&'_ GlobalCtxt<'_>>();
     |                             ^^^^^^^^^^^^^^^^^^ the trait `Hash` is not implemented for `rustc_ast::tokenstream::TokenStream`, which is required by `&GlobalCtxt<'_>: DynSync`
     |
     = note: required for `&rustc_ast::tokenstream::TokenStream` to implement `Hash`
     = note: 1 redundant requirement hidden
     = note: required for `(rustc_span::LocalExpnId, rustc_data_structures::svh::Svh, &rustc_ast::tokenstream::TokenStream)` to implement `Hash`
     = note: required for `DefaultCache<(LocalExpnId, Svh, &TokenStream), Erased<Erase<Result<&TokenStream, ()>>::{opaque#0}>>` to implement `rustc_query_system::query::QueryCache`
note: required because it appears within the type `DynamicQuery<'_, DefaultCache<(LocalExpnId, Svh, &TokenStream), Erased<Erase<Result<&TokenStream, ()>>::{opaque#0}>>>`
    --> compiler/rustc_middle/src/query/plumbing.rs:22:12
     |
22   | pub struct DynamicQuery<'tcx, C: QueryCache> {
     |            ^^^^^^^^^^^^
note: required because it appears within the type `DynamicQueries<'_>`
    --> compiler/rustc_middle/src/query/plumbing.rs:432:20
     |
275  |  / macro_rules! define_callbacks {
276  |  |     (
277  |  |      $($(#[$attr:meta])*
278  |  |         [$($modifiers:tt)*] fn $name:ident($($K:tt)*) -> $V:ty,)*) => {
...     |
432  |  |         pub struct DynamicQueries<'tcx> {
     |  |                    ^^^^^^^^^^^^^^
...     |
490  |  |     };
491  |  | }
     |  |_- in this expansion of `define_callbacks!` (#2)
     |
    ::: compiler/rustc_middle/src/query/mod.rs:108:1
     |
108  | /  rustc_queries! {
109  | |      query derive_macro_expansion(key: (LocalExpnId, Svh, &'tcx TokenStream)) -> Result<&'tcx TokenStream, ()> {
110  | |          // eval_always
111  | |          // no_hash
...    |
2335 | |      }
2336 | |  }
     | |  -
     | |  |
     | |__in this expansion of `rustc_query_append!` (#1)
     |    in this macro invocation (#2)
2337 |
2338 |    rustc_query_append! { define_callbacks! }
     |    ----------------------------------------- in this macro invocation (#1)
note: required because it appears within the type `QuerySystem<'_>`
    --> compiler/rustc_middle/src/query/plumbing.rs:59:12
     |
59   | pub struct QuerySystem<'tcx> {
     |            ^^^^^^^^^^^
note: required because it appears within the type `GlobalCtxt<'_>`
    --> compiler/rustc_middle/src/ty/context.rs:1280:12
     |
1280 | pub struct GlobalCtxt<'tcx> {
     |            ^^^^^^^^^^
     = note: required because it appears within the type `&GlobalCtxt<'_>`
note: required by a bound in `assert_dyn_sync`
    --> /home/felix/gits/rust/compiler/rustc_data_structures/src/marker.rs:192:36
     |
192  | pub fn assert_dyn_sync<T: ?Sized + DynSync>() {}
     |                                    ^^^^^^^ required by this bound in `assert_dyn_sync`
     = note: the full name for the type has been written to '/home/felix/gits/rust/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/rustc_middle-3afd8e69956fb57f.long-type-9950355256232114649.txt'
     = note: consider using `--verbose` to print the full type name to the console
     = note: the full name for the type has been written to '/home/felix/gits/rust/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/rustc_middle-3afd8e69956fb57f.long-type-14307373095868409618.txt'
     = note: consider using `--verbose` to print the full type name to the console

error[E0277]: the trait bound `rustc_ast::tokenstream::TokenStream: Hash` is not satisfied
    --> compiler/rustc_middle/src/ty/context.rs:1268:29
     |
1268 |     sync::assert_dyn_send::<&'_ GlobalCtxt<'_>>();
     |                             ^^^^^^^^^^^^^^^^^^ the trait `Hash` is not implemented for `rustc_ast::tokenstream::TokenStream`, which is required by `&GlobalCtxt<'_>: DynSend`
     |
     = note: required for `&rustc_ast::tokenstream::TokenStream` to implement `Hash`
     = note: 1 redundant requirement hidden
     = note: required for `(rustc_span::LocalExpnId, rustc_data_structures::svh::Svh, &rustc_ast::tokenstream::TokenStream)` to implement `Hash`
     = note: required for `DefaultCache<(LocalExpnId, Svh, &TokenStream), Erased<Erase<Result<&TokenStream, ()>>::{opaque#0}>>` to implement `rustc_query_system::query::QueryCache`
note: required because it appears within the type `DynamicQuery<'_, DefaultCache<(LocalExpnId, Svh, &TokenStream), Erased<Erase<Result<&TokenStream, ()>>::{opaque#0}>>>`
    --> compiler/rustc_middle/src/query/plumbing.rs:22:12
     |
22   | pub struct DynamicQuery<'tcx, C: QueryCache> {
     |            ^^^^^^^^^^^^
note: required because it appears within the type `DynamicQueries<'_>`
    --> compiler/rustc_middle/src/query/plumbing.rs:432:20
     |
275  |  / macro_rules! define_callbacks {
276  |  |     (
277  |  |      $($(#[$attr:meta])*
278  |  |         [$($modifiers:tt)*] fn $name:ident($($K:tt)*) -> $V:ty,)*) => {
...     |
432  |  |         pub struct DynamicQueries<'tcx> {
     |  |                    ^^^^^^^^^^^^^^
...     |
490  |  |     };
491  |  | }
     |  |_- in this expansion of `define_callbacks!` (#2)
     |
    ::: compiler/rustc_middle/src/query/mod.rs:108:1
     |
108  | /  rustc_queries! {
109  | |      query derive_macro_expansion(key: (LocalExpnId, Svh, &'tcx TokenStream)) -> Result<&'tcx TokenStream, ()> {
110  | |          // eval_always
111  | |          // no_hash
...    |
2335 | |      }
2336 | |  }
     | |  -
     | |  |
     | |__in this expansion of `rustc_query_append!` (#1)
     |    in this macro invocation (#2)
2337 |
2338 |    rustc_query_append! { define_callbacks! }
     |    ----------------------------------------- in this macro invocation (#1)
note: required because it appears within the type `QuerySystem<'_>`
    --> compiler/rustc_middle/src/query/plumbing.rs:59:12
     |
59   | pub struct QuerySystem<'tcx> {
     |            ^^^^^^^^^^^
note: required because it appears within the type `GlobalCtxt<'_>`
    --> compiler/rustc_middle/src/ty/context.rs:1280:12
     |
1280 | pub struct GlobalCtxt<'tcx> {
     |            ^^^^^^^^^^
     = note: required for `&GlobalCtxt<'_>` to implement `DynSend`
note: required by a bound in `assert_dyn_send`
    --> /home/felix/gits/rust/compiler/rustc_data_structures/src/marker.rs:193:36
     |
193  | pub fn assert_dyn_send<T: ?Sized + DynSend>() {}
     |                                    ^^^^^^^ required by this bound in `assert_dyn_send`
     = note: the full name for the type has been written to '/home/felix/gits/rust/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/rustc_middle-3afd8e69956fb57f.long-type-15445565275656164200.txt'
     = note: consider using `--verbose` to print the full type name to the console
     = note: the full name for the type has been written to '/home/felix/gits/rust/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/rustc_middle-3afd8e69956fb57f.long-type-2573109838703607371.txt'
     = note: consider using `--verbose` to print the full type name to the console

error[E0277]: the trait bound `rustc_ast::tokenstream::TokenStream: Hash` is not satisfied
    --> compiler/rustc_middle/src/query/plumbing.rs:386:21
     |
194  |            query_ensure($($args)*)
     |            ------------ required by a bound introduced by this call
...
275  |  / macro_rules! define_callbacks {
276  |  |     (
277  |  |      $($(#[$attr:meta])*
278  |  |         [$($modifiers:tt)*] fn $name:ident($($K:tt)*) -> $V:ty,)*) => {
...     |
386  |  |                     &self.tcx.query_system.caches.$name,
     |  |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Hash` is not implemented for `rustc_ast::tokenstream::TokenStream`, which is required by `DefaultCache<(rustc_span::LocalExpnId, rustc_data_structures::svh::Svh, &'tcx rustc_ast::tokenstream::TokenStream), Erased<query::erase::Erase<Result<&'tcx rustc_ast::tokenstream::TokenStream, ()>>::{opaque#0}>>: rustc_query_system::query::QueryCache`
...     |
490  |  |     };
491  |  | }
     |  |_- in this expansion of `define_callbacks!` (#2)
     |
    ::: compiler/rustc_middle/src/query/mod.rs:108:1
     |
108  | /  rustc_queries! {
109  | |      query derive_macro_expansion(key: (LocalExpnId, Svh, &'tcx TokenStream)) -> Result<&'tcx TokenStream, ()> {
110  | |          // eval_always
111  | |          // no_hash
...    |
2335 | |      }
2336 | |  }
     | |  -
     | |  |
     | |__in this expansion of `rustc_query_append!` (#1)
     |    in this macro invocation (#2)
2337 |
2338 |    rustc_query_append! { define_callbacks! }
     |    ----------------------------------------- in this macro invocation (#1)
     |
     = help: the trait `rustc_query_system::query::QueryCache` is implemented for `DefaultCache<K, V>`
     = note: required for `&'tcx rustc_ast::tokenstream::TokenStream` to implement `Hash`
     = note: 1 redundant requirement hidden
     = note: required for `(rustc_span::LocalExpnId, rustc_data_structures::svh::Svh, &'tcx rustc_ast::tokenstream::TokenStream)` to implement `Hash`
     = note: required for `DefaultCache<(LocalExpnId, Svh, &TokenStream), Erased<Erase<Result<&TokenStream, ()>>::{opaque#0}>>` to implement `rustc_query_system::query::QueryCache`
note: required by a bound in `query_ensure`
    --> compiler/rustc_middle/src/query/plumbing.rs:155:12
     |
148  | pub fn query_ensure<'tcx, Cache>(
     |        ------------ required by a bound in this function
...
155  |     Cache: QueryCache,
     |            ^^^^^^^^^^ required by this bound in `query_ensure`
     = note: the full name for the type has been written to '/home/felix/gits/rust/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/rustc_middle-3afd8e69956fb57f.long-type-16771981020465904745.txt'
     = note: consider using `--verbose` to print the full type name to the console

error[E0277]: the trait bound `rustc_ast::tokenstream::TokenStream: Hash` is not satisfied
    --> compiler/rustc_middle/src/query/plumbing.rs:194:9
     |
192  |      / macro_rules! query_ensure {
193  |      |     ([]$($args:tt)*) => {
194  |      |         query_ensure($($args)*)
     |      |         ^^^^^^^^^^^^^^^^^^^^^^^ the trait `Hash` is not implemented for `rustc_ast::tokenstream::TokenStream`, which is required by `DefaultCache<(rustc_span::LocalExpnId, rustc_data_structures::svh::Svh, &'tcx rustc_ast::tokenstream::TokenStream), Erased<query::erase::Erase<Result<&'tcx rustc_ast::tokenstream::TokenStream, ()>>::{opaque#0}>>: rustc_query_system::query::QueryCache`
195  |      |     };
...         |
200  |      |         query_ensure!([$($modifiers)*]$($args)*)
     |      |         ----------------------------------------
     |      |         |
     |      |         in this macro invocation (#4)
     |      |         in this macro invocation (#5)
201  |      |     };
202  |      | }
     |      | -
     |      | |
     |      | in this expansion of `query_ensure!` (#3)
     |      |_in this expansion of `query_ensure!` (#4)
     |        in this expansion of `query_ensure!` (#5)
...
275  |      / macro_rules! define_callbacks {
276  |      |     (
277  |      |      $($(#[$attr:meta])*
278  |      |         [$($modifiers:tt)*] fn $name:ident($($K:tt)*) -> $V:ty,)*) => {
...         |
382  |      /                 query_ensure!(
383  |                            [$($modifiers)*]
384  |                            self.tcx,
385  |                            self.tcx.query_system.fns.engine.$name,
...
388  |                            false,
389  |      |                 )
     |      |_________________- in this macro invocation (#3)
...
490  |      |     };
491  |      | }
     |      |_- in this expansion of `define_callbacks!` (#2)
     |
    ::: compiler/rustc_middle/src/query/mod.rs:108:1
     |
108  | /      rustc_queries! {
109  | |          query derive_macro_expansion(key: (LocalExpnId, Svh, &'tcx TokenStream)) -> Result<&'tcx TokenStream, ()> {
110  | |              // eval_always
111  | |              // no_hash
...    |
2335 | |          }
2336 | |      }
     | |      -
     | |      |
     | |______in this expansion of `rustc_query_append!` (#1)
     |        in this macro invocation (#2)
2337 |
2338 |        rustc_query_append! { define_callbacks! }
     |        ----------------------------------------- in this macro invocation (#1)
     |
     = help: the trait `rustc_query_system::query::QueryCache` is implemented for `DefaultCache<K, V>`
     = note: required for `&'tcx rustc_ast::tokenstream::TokenStream` to implement `Hash`
     = note: 1 redundant requirement hidden
     = note: required for `(rustc_span::LocalExpnId, rustc_data_structures::svh::Svh, &'tcx rustc_ast::tokenstream::TokenStream)` to implement `Hash`
     = note: required for `DefaultCache<(LocalExpnId, Svh, &TokenStream), Erased<Erase<Result<&TokenStream, ()>>::{opaque#0}>>` to implement `rustc_query_system::query::QueryCache`
     = note: the full name for the type has been written to '/home/felix/gits/rust/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/rustc_middle-3afd8e69956fb57f.long-type-16771981020465904745.txt'
     = note: consider using `--verbose` to print the full type name to the console

error[E0277]: the trait bound `rustc_ast::tokenstream::TokenStream: Hash` is not satisfied
    --> compiler/rustc_middle/src/query/plumbing.rs:400:21
     |
275  |  / macro_rules! define_callbacks {
276  |  |     (
277  |  |      $($(#[$attr:meta])*
278  |  |         [$($modifiers:tt)*] fn $name:ident($($K:tt)*) -> $V:ty,)*) => {
...     |
397  |  |                 query_ensure(
     |  |                 ------------ required by a bound introduced by this call
...     |
400  |  |                     &self.tcx.query_system.caches.$name,
     |  |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Hash` is not implemented for `rustc_ast::tokenstream::TokenStream`, which is required by `DefaultCache<(rustc_span::LocalExpnId, rustc_data_structures::svh::Svh, &'tcx rustc_ast::tokenstream::TokenStream), Erased<query::erase::Erase<Result<&'tcx rustc_ast::tokenstream::TokenStream, ()>>::{opaque#0}>>: rustc_query_system::query::QueryCache`
...     |
490  |  |     };
491  |  | }
     |  |_- in this expansion of `define_callbacks!` (#2)
     |
    ::: compiler/rustc_middle/src/query/mod.rs:108:1
     |
108  | /  rustc_queries! {
109  | |      query derive_macro_expansion(key: (LocalExpnId, Svh, &'tcx TokenStream)) -> Result<&'tcx TokenStream, ()> {
110  | |          // eval_always
111  | |          // no_hash
...    |
2335 | |      }
2336 | |  }
     | |  -
     | |  |
     | |__in this expansion of `rustc_query_append!` (#1)
     |    in this macro invocation (#2)
2337 |
2338 |    rustc_query_append! { define_callbacks! }
     |    ----------------------------------------- in this macro invocation (#1)
     |
     = help: the trait `rustc_query_system::query::QueryCache` is implemented for `DefaultCache<K, V>`
     = note: required for `&'tcx rustc_ast::tokenstream::TokenStream` to implement `Hash`
     = note: 1 redundant requirement hidden
     = note: required for `(rustc_span::LocalExpnId, rustc_data_structures::svh::Svh, &'tcx rustc_ast::tokenstream::TokenStream)` to implement `Hash`
     = note: required for `DefaultCache<(LocalExpnId, Svh, &TokenStream), Erased<Erase<Result<&TokenStream, ()>>::{opaque#0}>>` to implement `rustc_query_system::query::QueryCache`
note: required by a bound in `query_ensure`
    --> compiler/rustc_middle/src/query/plumbing.rs:155:12
     |
148  | pub fn query_ensure<'tcx, Cache>(
     |        ------------ required by a bound in this function
...
155  |     Cache: QueryCache,
     |            ^^^^^^^^^^ required by this bound in `query_ensure`
     = note: the full name for the type has been written to '/home/felix/gits/rust/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/rustc_middle-3afd8e69956fb57f.long-type-16771981020465904745.txt'
     = note: consider using `--verbose` to print the full type name to the console

error[E0277]: the trait bound `rustc_ast::tokenstream::TokenStream: Hash` is not satisfied
    --> compiler/rustc_middle/src/query/plumbing.rs:397:17
     |
275  |  /  macro_rules! define_callbacks {
276  |  |      (
277  |  |       $($(#[$attr:meta])*
278  |  |          [$($modifiers:tt)*] fn $name:ident($($K:tt)*) -> $V:ty,)*) => {
...     |
397  |  |/                 query_ensure(
398  |  ||                     self.tcx,
399  |  ||                     self.tcx.query_system.fns.engine.$name,
400  |  ||                     &self.tcx.query_system.caches.$name,
401  |  ||                     key.into_query_param(),
402  |  ||                     true,
403  |  ||                 );
     |  ||_________________^ the trait `Hash` is not implemented for `rustc_ast::tokenstream::TokenStream`, which is required by `DefaultCache<(rustc_span::LocalExpnId, rustc_data_structures::svh::Svh, &'tcx rustc_ast::tokenstream::TokenStream), Erased<query::erase::Erase<Result<&'tcx rustc_ast::tokenstream::TokenStream, ()>>::{opaque#0}>>: rustc_query_system::query::QueryCache`
...     |
490  |  |      };
491  |  |  }
     |  |__- in this expansion of `define_callbacks!` (#2)
     |
    ::: compiler/rustc_middle/src/query/mod.rs:108:1
     |
108  | /   rustc_queries! {
109  | |       query derive_macro_expansion(key: (LocalExpnId, Svh, &'tcx TokenStream)) -> Result<&'tcx TokenStream, ()> {
110  | |           // eval_always
111  | |           // no_hash
...    |
2335 | |       }
2336 | |   }
     | |   -
     | |   |
     | |___in this expansion of `rustc_query_append!` (#1)
     |     in this macro invocation (#2)
2337 |
2338 |     rustc_query_append! { define_callbacks! }
     |     ----------------------------------------- in this macro invocation (#1)
     |
     = help: the trait `rustc_query_system::query::QueryCache` is implemented for `DefaultCache<K, V>`
     = note: required for `&'tcx rustc_ast::tokenstream::TokenStream` to implement `Hash`
     = note: 1 redundant requirement hidden
     = note: required for `(rustc_span::LocalExpnId, rustc_data_structures::svh::Svh, &'tcx rustc_ast::tokenstream::TokenStream)` to implement `Hash`
     = note: required for `DefaultCache<(LocalExpnId, Svh, &TokenStream), Erased<Erase<Result<&TokenStream, ()>>::{opaque#0}>>` to implement `rustc_query_system::query::QueryCache`
     = note: the full name for the type has been written to '/home/felix/gits/rust/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/rustc_middle-3afd8e69956fb57f.long-type-16771981020465904745.txt'
     = note: consider using `--verbose` to print the full type name to the console

error[E0277]: the trait bound `rustc_ast::tokenstream::TokenStream: Hash` is not satisfied
    --> compiler/rustc_middle/src/query/plumbing.rs:425:21
     |
275  |  / macro_rules! define_callbacks {
276  |  |     (
277  |  |      $($(#[$attr:meta])*
278  |  |         [$($modifiers:tt)*] fn $name:ident($($K:tt)*) -> $V:ty,)*) => {
...     |
422  |  |                 restore::<$V>(query_get_at(
     |  |                               ------------ required by a bound introduced by this call
...     |
425  |  |                     &self.tcx.query_system.caches.$name,
     |  |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Hash` is not implemented for `rustc_ast::tokenstream::TokenStream`, which is required by `DefaultCache<(rustc_span::LocalExpnId, rustc_data_structures::svh::Svh, &'tcx rustc_ast::tokenstream::TokenStream), Erased<query::erase::Erase<Result<&'tcx rustc_ast::tokenstream::TokenStream, ()>>::{opaque#0}>>: rustc_query_system::query::QueryCache`
...     |
490  |  |     };
491  |  | }
     |  |_- in this expansion of `define_callbacks!` (#2)
     |
    ::: compiler/rustc_middle/src/query/mod.rs:108:1
     |
108  | /  rustc_queries! {
109  | |      query derive_macro_expansion(key: (LocalExpnId, Svh, &'tcx TokenStream)) -> Result<&'tcx TokenStream, ()> {
110  | |          // eval_always
111  | |          // no_hash
...    |
2335 | |      }
2336 | |  }
     | |  -
     | |  |
     | |__in this expansion of `rustc_query_append!` (#1)
     |    in this macro invocation (#2)
2337 |
2338 |    rustc_query_append! { define_callbacks! }
     |    ----------------------------------------- in this macro invocation (#1)
     |
     = help: the trait `rustc_query_system::query::QueryCache` is implemented for `DefaultCache<K, V>`
     = note: required for `&'tcx rustc_ast::tokenstream::TokenStream` to implement `Hash`
     = note: 1 redundant requirement hidden
     = note: required for `(rustc_span::LocalExpnId, rustc_data_structures::svh::Svh, &'tcx rustc_ast::tokenstream::TokenStream)` to implement `Hash`
     = note: required for `DefaultCache<(LocalExpnId, Svh, &TokenStream), Erased<Erase<Result<&TokenStream, ()>>::{opaque#0}>>` to implement `rustc_query_system::query::QueryCache`
note: required by a bound in `query_get_at`
    --> compiler/rustc_middle/src/query/plumbing.rs:138:12
     |
130  | pub fn query_get_at<'tcx, Cache>(
     |        ------------ required by a bound in this function
...
138  |     Cache: QueryCache,
     |            ^^^^^^^^^^ required by this bound in `query_get_at`
     = note: the full name for the type has been written to '/home/felix/gits/rust/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/rustc_middle-3afd8e69956fb57f.long-type-16771981020465904745.txt'
     = note: consider using `--verbose` to print the full type name to the console

error[E0277]: the trait bound `rustc_ast::tokenstream::TokenStream: Hash` is not satisfied
    --> compiler/rustc_middle/src/query/plumbing.rs:422:31
     |
275  |  /  macro_rules! define_callbacks {
276  |  |      (
277  |  |       $($(#[$attr:meta])*
278  |  |          [$($modifiers:tt)*] fn $name:ident($($K:tt)*) -> $V:ty,)*) => {
...     |
422  |  |                  restore::<$V>(query_get_at(
     |  | _______________________________^
423  |  ||                     self.tcx,
424  |  ||                     self.tcx.query_system.fns.engine.$name,
425  |  ||                     &self.tcx.query_system.caches.$name,
426  |  ||                     self.span,
427  |  ||                     key.into_query_param(),
428  |  ||                 ))
     |  ||_________________^ the trait `Hash` is not implemented for `rustc_ast::tokenstream::TokenStream`, which is required by `DefaultCache<(rustc_span::LocalExpnId, rustc_data_structures::svh::Svh, &'tcx rustc_ast::tokenstream::TokenStream), Erased<query::erase::Erase<Result<&'tcx rustc_ast::tokenstream::TokenStream, ()>>::{opaque#0}>>: rustc_query_system::query::QueryCache`
...     |
490  |  |      };
491  |  |  }
     |  |__- in this expansion of `define_callbacks!` (#2)
     |
    ::: compiler/rustc_middle/src/query/mod.rs:108:1
     |
108  | /   rustc_queries! {
109  | |       query derive_macro_expansion(key: (LocalExpnId, Svh, &'tcx TokenStream)) -> Result<&'tcx TokenStream, ()> {
110  | |           // eval_always
111  | |           // no_hash
...    |
2335 | |       }
2336 | |   }
     | |   -
     | |   |
     | |___in this expansion of `rustc_query_append!` (#1)
     |     in this macro invocation (#2)
2337 |
2338 |     rustc_query_append! { define_callbacks! }
     |     ----------------------------------------- in this macro invocation (#1)
     |
     = help: the trait `rustc_query_system::query::QueryCache` is implemented for `DefaultCache<K, V>`
     = note: required for `&'tcx rustc_ast::tokenstream::TokenStream` to implement `Hash`
     = note: 1 redundant requirement hidden
     = note: required for `(rustc_span::LocalExpnId, rustc_data_structures::svh::Svh, &'tcx rustc_ast::tokenstream::TokenStream)` to implement `Hash`
     = note: required for `DefaultCache<(LocalExpnId, Svh, &TokenStream), Erased<Erase<Result<&TokenStream, ()>>::{opaque#0}>>` to implement `rustc_query_system::query::QueryCache`
     = note: the full name for the type has been written to '/home/felix/gits/rust/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/rustc_middle-3afd8e69956fb57f.long-type-16771981020465904745.txt'
     = note: consider using `--verbose` to print the full type name to the console

For more information about this error, try `rustc --explain E0277`.
error: could not compile `rustc_middle` (lib) due to 9 previous errors

</pre>
</details>